### PR TITLE
Allow multiple serializer instances

### PR DESCRIPTION
**Pull request by goetas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Doc updated   | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT

Allows to do something as:

```yaml
jms_serializer:
    metadata:
        auto_detection: false
        directories:
            entity:
                namespace_prefix: 'App\Entity'
                path: "%kernel.project_dir%/src/App/Resources/metadata/entity"

    instances:

        serializer_2: # same instance as the default one, but with different naming strategy
            inherit: true
            property_naming:
                separator: '-'
                lower_case: false

        serializer_3: # different instance loading different metadata from the same entities
            inherit: false
            metadata:
                auto_detection: false
                directories:
                    entity:
                        namespace_prefix: 'App\Entity'
                        path: "%kernel.project_dir%/src/App/Resources/another-metadata/entity"

```

This will provide 3 different serializer instances:
- `jms_serializer.instances.default` (or `jms_seralizer`) with the main configs as before
- `jms_serializer.instances.serializer_2`, same configs as `jms_seralizer` but with a different naming strategy
- `jms_serializer.instances.serializer_3` a brand new serializer instance with a different metadata source

It is also possible to define global and per instance event and type handlers/subscribers/listeners.

Rework of https://github.com/schmittjoh/JMSSerializerBundle/pull/831

---------------

I have to admit that this feature, very cool (and useful) in my opinion, will be very hard to maintain in the future! It does a lot of woodoo with the symfony dependency injection container! I'm almost afraid to merge it :scream: !

### CODE REVIEW

File: Debug/DataCollector.php [hunk 1]
```diff
@@ -51,11 +54,21 @@ public function reset(): void
...
+
+    public function getInstance(): string
+    {
+        return $this->data['instance'];
     }
 
     public function getName(): string
     {
-        return 'jms_serializer';
+        if (($this->instance ?? $this->data['instance']) === 'default'){
```
Comment (scyzoryck) [inline Debug/DataCollector.php:67]: Looking at constructor - `$this->instance` is not nullable so `??` seems to be redundant.

File: DependencyInjection/DIUtils.php [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+declare(strict_types=1);
+
+namespace JMS\SerializerBundle\DependencyInjection;
+
+use Symfony\Component\DependencyInjection\Alias;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+
+ini_set('xdebug.overload_var_dump', 'off');
```
Comment (scyzoryck) [inline DependencyInjection/DIUtils.php:12]: Looks like some leftover from debuging :D

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -35,17 +38,47 @@ public function getConfigTreeBuilder()
...
         $this->addContextSection($root);
+    }
 
-        return $tb;
+    private function addProfilerSection(NodeBuilder $builder): void
+    {
+        $builder->scalarNode('profiler')
+            ->defaultValue(null)
+            ->validate()
+            ->always(static function ($v) {
```
Comment (scyzoryck) [inline DependencyInjection/Configuration.php:73]:
```suggestion
            ->always(static function ($v): ?bool {
```

File: DependencyInjection/ScopedContainer.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                if (empty($attributes['instance']) || $attributes['instance'] === $this->instance) {
+                    $serviceIds[$id][] = $attributes;
+                }
+            }
+        }
+
+        return $serviceIds;
+    }
+
+    private function getDefinitionRealId($id): string
```
Comment (scyzoryck) [inline DependencyInjection/ScopedContainer.php:65]: `getRealId` requires string as second parameter, so it should be also required here I guess.
```suggestion
    private function getDefinitionRealId(string $id): string
```

File: Debug/DataCollector.php (hunk 1 above)
Comment (goetas) [reply Debug/DataCollector.php:67]: yea, this is something that I still need to figure out. there is something different when the data collector is created and used in the profiler versus when is used to collect the data while running the application.

i will look better into this

**Comment by tarjei** [comment]
This is looking good! One question, is there a way when we are mangling with Symfony DI to add bind aliases or should that be left to the application?

**Comment by goetas** [comment]
if you bind `jms_serializer.instances.serializer_2` it will take the instance `serializer_2`. if you bind `Jms\Serializer\SerializerInterface` it will take the }default` instance. The only type of configuration is to allow binding `Jms\Serializer\SerializerInterface` to a custom instance instead of a default one. Is that what you mean?

**Comment by goetas** [comment]
done as part of https://github.com/schmittjoh/JMSSerializerBundle/pull/906

**Pull request closed without merging**